tf_encrypted/protocol/securenn/securenn.py

```diff
@@ -353,9 +353,38 @@ def relu(self, x):
 
     :param PondTensor x: Input tensor.
     """
-    with tf.name_scope('relu'):
-      drelu = self.non_negative(x)
-      return drelu * x
+
+    def relu(x):
+      with tf.name_scope('relu'):
+        drelu = self.non_negative(x)
+        return drelu * x
+
+
+    max_size = 224*224*95
+    shape = x.shape.as_list()
+
+    if np.prod(shape) >= max_size:
+      #tensor is too big and will raise OOM
+
+      def check_num_split(num_split):
+        return (shape[-1] % num_split == 0 and
+                np.prod(shape[:-1] + [shape[-1] / num_split]) < max_size)
+
+
+      #look for best num_split value
+      num_split = 2
+      while not check_num_split(num_split):
```

Comment:
looping like this might be a bit expensive, but since it all happens at compile time it might not be a big issue. however, i believe it could cause a livelock in the cases where it is not enough to only split the last dimension, ie if `shape[:-1] + [1]` is still too large.

an alternative approach would be to not insist on splitting into equal sized subtensors but instead take a number of maxsize tensors as possible and a single left-over tensor.

in any case, taking all dimensions into account is probably overkill for now, and think it's fine to only consider how to split the last dimension for now. it should raise an error instead of livelocking though.

Reply:
I was considering flattening the tensor, splitting as you suggested, performing ReLU and then concatenating and reshaping to original shape. This way we don't limit the implementation to the last dimension, but this also adds the cost of the flattening and reshaping. @mortendahl What do you think?

Reply:
I would go for the simple solution for now and only split along the last dimension (passing a vector to `tf.split`), but raise an error in the cases where that's not enough.

If you prefer the fuller solution then I think flattening and reshaping is actually very cheap, as I believe they only change the view of the data, not how it's laid out in memory. Splitting and joining is the more expensive operations as they might copy data around.

Reply:
I'm not sure if this helps much, but `tf.split` accepts a 1d-array for the `axis` argument as well as the `num_size` argument, which allows it to split across multiple dimensions at once